### PR TITLE
GH-44066: [Python] Add Python wrapper for JsonExtensionType

### PR DESCRIPTION
**Pull request by rok**
### Rationale for this change

We [added canonical JsonExtensionType](https://github.com/apache/arrow/pull/13901) and we should make it usable from Python.

### What changes are included in this PR?

Python wrapper for `JsonExtensionType` and `JsonArray` are added on Python side as well as `JsonArray` on c++ side.

### Are these changes tested?

Python tests for the extension type and array are included.

### Are there any user-facing changes?

This adds a json canonical extension type to pyarrow.
* GitHub Issue: #44066

### CODE REVIEW

File: python/pyarrow/array.pxi
Comment (jorisvandenbossche) [inline python/pyarrow/array.pxi:4349]: Should we maybe note here that this array class does not guarantee that the data is actually valid JSON? (which I assume is the case?)

For example in the code example below, you can in theory put whatever string data you want in the array.

File: python/pyarrow/lib.pxd
Comment (jorisvandenbossche) [inline python/pyarrow/lib.pxd:232]:
```suggestion


```

File: cpp/src/arrow/extension/json.h
Comment (jorisvandenbossche) [inline cpp/src/arrow/extension/json.h:33]: This is not used in C++, but I think also not actually needed for the Python bindings?

File: python/pyarrow/types.pxi
Comment (jorisvandenbossche) [inline python/pyarrow/types.pxi:5276]: Use a default of utf8 ?

File: python/pyarrow/tests/test_extension_type.py
Comment (jorisvandenbossche) [inline python/pyarrow/tests/test_extension_type.py:1931]: The test doesn't work yet for `pa.utf8_view()`?

File: python/pyarrow/tests/test_extension_type.py
Comment (jorisvandenbossche) [inline python/pyarrow/tests/test_extension_type.py:1940]: Can you also check equality at the type instance level? 

```suggestion
    assert json_type.__class__ is pa.JsonType

    assert json_type == pa.json(storage_type)
    assert json_type != storage_type
```

Which also makes me wonder: if the storage type is different (eg string vs large string), does it then return False? (maybe also good to test that)

File: python/pyarrow/tests/test_extension_type.py
Comment (jorisvandenbossche) [inline python/pyarrow/tests/test_extension_type.py:1948]: Can you also add a pickle roundtrip of the type? (see eg test_opaque_type above)

Comment (jorisvandenbossche) [reply python/pyarrow/tests/test_extension_type.py:1948]: And maybe also a basic cast (again see the structure of `test_opaque_type` and `test_bool8_type` above, you can follow that I think)

**Review by jorisvandenbossche — COMMENTED** [review]
Thanks for working on this!

Added some inline comments. One other comment:

- Can you add the new classes to `test_extension_type_constructor_errors` in test_misc.py

Are you planning to work on enabling the parquet integration in a later PR?

File: python/pyarrow/array.pxi
Comment (rok) [reply python/pyarrow/array.pxi:4351]: Added a not. Let me know if it's not extensive enough.

File: cpp/src/arrow/extension/json.h
Comment (rok) [reply cpp/src/arrow/extension/json.h]: Indeed it is not. Removed.

File: python/pyarrow/tests/test_extension_type.py
Comment (rok) [reply python/pyarrow/tests/test_extension_type.py]: It does, expanded the test set.

File: python/pyarrow/tests/test_extension_type.py
Comment (rok) [reply python/pyarrow/tests/test_extension_type.py]: Added pickle and cast tests. Cast test fails for `extenstion.cast(string_view)`. I've not checked why yet but probably we're missing a cast kernel for view types?
```python
pyarrow.lib.ArrowNotImplementedError: Unsupported cast to string_view from extension<arrow.json>
```

**Comment by rok** [comment]
> Thanks for working on this!

Thanks for the review!

> * Can you add the new classes to `test_extension_type_constructor_errors` in test_misc.py

Added.

> Are you planning to work on enabling the parquet integration in a later PR?

Do we need to do something Python-side? C++ side was covered by #13901. I've added a basic parquet test here and it works for me locally.

Question: any idea why pytest is saying `pa.json` is [not callable in CI](https://github.com/apache/arrow/actions/runs/10828781440/job/30044921830?pr=44070#step:6:6727)? Locally this works as expected.

File: python/pyarrow/tests/test_extension_type.py
Comment (jorisvandenbossche) [reply python/pyarrow/tests/test_extension_type.py]: Yes, I think casting for string view is not yet implemented

**Comment by jorisvandenbossche** [comment]
> Do we need to do something Python-side? C++ side was covered by #13901. I've added a basic parquet test here and it works for me locally.

I would be surprised if it works locally out of the box because IIRC the option was disabled by default in C++ for now? (so what we would need on the python side is expose that new `arrow_extensions_enabled` options so that the user can enable it) 
Although I suppose that you wrote the file with pyarrow, and at that point we give priority to the stored schema, I suppose, even if the option is not enabled.



> Question: any idea why pytest is saying `pa.json` is [not callable in CI](https://github.com/apache/arrow/actions/runs/10828781440/job/30044921830?pr=44070#step:6:6727)? Locally this works as expected.

There might somehow we some conflict with the stdlib module?

**Comment by jorisvandenbossche** [comment]
> > Question: any idea why pytest is saying `pa.json` is [not callable in CI](https://github.com/apache/arrow/actions/runs/10828781440/job/30044921830?pr=44070#step:6:6727)? Locally this works as expected.
> 
> There might somehow we some conflict with the stdlib module?

Actually not that, but we have a `pyarrow.json` module already. So we will have to call this function differently anyway. For others with a conflict with added a trailing underscore, so could do that here as well.

File: python/pyarrow/types.pxi
Comment (rok) [reply python/pyarrow/types.pxi]: Done.

File: python/pyarrow/tests/test_extension_type.py
Comment (rok) [reply python/pyarrow/tests/test_extension_type.py]: I suppose #39634 covers casting.

File: python/pyarrow/tests/test_extension_type.py
Comment (rok) [reply python/pyarrow/tests/test_extension_type.py]: Moved.

**Comment by rok** [comment]
> > Do we need to do something Python-side? C++ side was covered by #13901. I've added a basic parquet test here and it works for me locally.
> 
> I would be surprised if it works locally out of the box because IIRC the option was disabled by default in C++ for now? (so what we would need on the python side is expose that new `arrow_extensions_enabled` options so that the user can enable it) Although I suppose that you wrote the file with pyarrow, and at that point we give priority to the stored schema, I suppose, even if the option is not enabled.

I see. I assumed `arrow_extensions_enabled` was on by default. Do we want to expose a global setter? E.g.: `pyarrow.parquet.set_arrow_extensions_enabled`? Or would it be preferred to add a parameter for `pq.write_table`?

> > Question: any idea why pytest is saying `pa.json` is [not callable in CI](https://github.com/apache/arrow/actions/runs/10828781440/job/30044921830?pr=44070#step:6:6727)? Locally this works as expected.
> 
> There might somehow we some conflict with the stdlib module?
> [..]
> Actually not that, but we have a pyarrow.json module already. So we will have to call this function differently anyway. For > others with a conflict with added a trailing underscore, so could do that here as well.

It seems this was the case, changed extension type name to `pa.json_`.

File: python/pyarrow/tests/test_extension_type.py
Comment (felipecrv) [inline python/pyarrow/tests/test_extension_type.py:1975]: Nice!

Comment (rok) [reply python/pyarrow/tests/test_extension_type.py:1975]: Thanks for providing that :)

**Comment by rok** [comment]
@jorisvandenbossche ping :)

File: python/pyarrow/array.pxi
Comment (jorisvandenbossche) [inline python/pyarrow/array.pxi:4349]:
```suggestion
    Concrete class for Arrow arrays of JSON data type.

    This does not guarantee that the JSON data actually
```

(blank line between first summary line and the extended explanation)

File: python/pyarrow/array.pxi
Comment (jorisvandenbossche) [inline python/pyarrow/array.pxi:4350]:
```suggestion
    is valid JSON.
```

(missing verb?)

File: python/pyarrow/scalar.pxi
Comment (jorisvandenbossche) [inline python/pyarrow/scalar.pxi:1053]: Is this actually needed? (that's what the base class already does?

File: python/pyarrow/tests/parquet/test_data_types.py
Comment (jorisvandenbossche) [inline python/pyarrow/tests/parquet/test_data_types.py:516]:
```suggestion
    pa.string(), pa.large_string()))
```

No need to parametrize with both ut8 and string, as they are just aliases

File: python/pyarrow/tests/parquet/test_data_types.py
Comment (jorisvandenbossche) [inline python/pyarrow/tests/parquet/test_data_types.py:525]: Does it not work with use_dictionary=True ?

File: python/pyarrow/tests/test_extension_type.py
Comment (jorisvandenbossche) [inline python/pyarrow/tests/test_extension_type.py:1932]:
```suggestion
    pa.string(), pa.large_string(), pa.string_view()))
```

File: python/pyarrow/tests/test_extension_type.py
Comment (jorisvandenbossche) [inline python/pyarrow/tests/test_extension_type.py:1939]: So this means that the storage type is not taken into account for checking equality of the JSON extension type? Should it?

File: python/pyarrow/types.pxi
Comment (jorisvandenbossche) [inline python/pyarrow/types.pxi:1779]: This is not the class for the array but for the type

File: python/pyarrow/types.pxi
Comment (jorisvandenbossche) [inline python/pyarrow/types.pxi:5282]:
```suggestion
    storage_type : DataType, default pyarrow.string()
```

File: python/pyarrow/types.pxi
Comment (jorisvandenbossche) [inline python/pyarrow/types.pxi:5283]: Maybe mention here that this should be one of the string data types? (also, do you get a proper error message when passing a non-string data type? Can you test that?)

**Review by jorisvandenbossche — COMMENTED** [review]
Thanks for the update! Some more comments

File: python/pyarrow/tests/test_extension_type.py
Comment (rok) [reply python/pyarrow/tests/test_extension_type.py:1939]: I meant to test that `pa.json_()` sets storage to `pa.utf8()` by default and I'd keep this test. However as noted in https://github.com/apache/arrow/pull/13901#pullrequestreview-2324294761 disregarding storage type here is not desirable.

File: python/pyarrow/types.pxi
Comment (rok) [reply python/pyarrow/types.pxi]: Changed to reflect that.

File: python/pyarrow/tests/parquet/test_data_types.py
Comment (rok) [reply python/pyarrow/tests/parquet/test_data_types.py]: It does. Removed `use_dictionary=False`.

File: python/pyarrow/tests/test_extension_type.py
Comment (jorisvandenbossche) [reply python/pyarrow/tests/test_extension_type.py:1939]: > I meant to test that `pa.json_()` sets storage to `pa.utf8()` by default and I'd keep this test.

Ah, yes of course, that's good to test

File: python/pyarrow/tests/parquet/test_data_types.py
Comment (jorisvandenbossche) [inline python/pyarrow/tests/parquet/test_data_types.py:531]: Can you move this to the test about the extension type itself? (this is not related to parquet)

Comment (rok) [reply python/pyarrow/tests/parquet/test_data_types.py]: Oh, sorry. Moved.

**Comment by rok** [comment]
@jorisvandenbossche I've rebased this on https://github.com/apache/arrow/pull/44215, which is now semi-approved. So if you can spare the time review would be greatly appreciated :D.

**Comment by rok** [comment]
Another rebase on master.

**Comment by rok** [comment]
@pitrou could you do a quick pass here in case anything stands out please?

File: python/pyarrow/types.pxi
Comment (pitrou) [inline python/pyarrow/types.pxi:1829]: Side note: it would be nice if one could write `json_array = pa.array(arr, json_type)`.
Perhaps open a feature request?

File: python/pyarrow/types.pxi
Comment (pitrou) [inline python/pyarrow/types.pxi:5345]: One superfluous line here.

File: python/pyarrow/tests/test_extension_type.py
Comment (pitrou) [inline python/pyarrow/tests/test_extension_type.py:1979]: Pity this doesn't raise `TypeError`.

File: python/pyarrow/tests/test_extension_type.py
Comment (pitrou) [inline python/pyarrow/tests/test_extension_type.py:1981]: Nit: can use a f-string

File: python/pyarrow/types.pxi
Comment (rok) [reply python/pyarrow/types.pxi:1829]: That would be a nice feature. https://github.com/apache/arrow/issues/44406

File: python/pyarrow/types.pxi
Comment (rok) [reply python/pyarrow/types.pxi:5345]: Removed.

File: python/pyarrow/tests/test_extension_type.py
Comment (rok) [reply python/pyarrow/tests/test_extension_type.py:1979]: We could catch and raise it but it's probably not a good idea.

File: python/pyarrow/tests/test_extension_type.py
Comment (rok) [reply python/pyarrow/tests/test_extension_type.py]: Done.

**Comment by rok** [comment]
@pitrou do you think this merits a merge or should we wait for another review?

**Comment by pitrou** [comment]
What's the plan for the Parquet  `arrow_extensions_enabled` option?

**Comment by rok** [comment]
@pitrou
> What's the plan for the Parquet `arrow_extensions_enabled` option?

Perhaps we should open another issue for it? Current implementation seems to roundtrip to parquet ok.
I'd propose something like this:

```diff
diff --git a/python/pyarrow/_parquet.pxd b/python/pyarrow/_parquet.pxd
index d6aebd8284..32e2618ecf 100644
--- a/python/pyarrow/_parquet.pxd
+++ b/python/pyarrow/_parquet.pxd
@@ -405,6 +405,7 @@ cdef extern from "parquet/api/reader.h" namespace "parquet" nogil:
         CCacheOptions cache_options() const
         void set_coerce_int96_timestamp_unit(TimeUnit unit)
         TimeUnit coerce_int96_timestamp_unit() const
+        void set_arrow_extensions_enabled(c_bool enabled)
 
     ArrowReaderProperties default_arrow_reader_properties()
 
diff --git a/python/pyarrow/_parquet.pyx b/python/pyarrow/_parquet.pyx
index 254bfe3b09..6ae1726c71 100644
--- a/python/pyarrow/_parquet.pyx
+++ b/python/pyarrow/_parquet.pyx
@@ -1441,7 +1441,8 @@ cdef class ParquetReader(_Weakrefable):
              FileDecryptionProperties decryption_properties=None,
              thrift_string_size_limit=None,
              thrift_container_size_limit=None,
-             page_checksum_verification=False):
+             page_checksum_verification=False,
+             arrow_extensions_enabled=False):
         """
         Open a parquet file for reading.
 
@@ -1458,6 +1459,7 @@ cdef class ParquetReader(_Weakrefable):
         thrift_string_size_limit : int, optional
         thrift_container_size_limit : int, optional
         page_checksum_verification : bool, default False
+        arrow_extensions_enabled: bool, default False
         """
         cdef:
             shared_ptr[CFileMetaData] c_metadata
@@ -1522,6 +1524,9 @@ cdef class ParquetReader(_Weakrefable):
         if read_dictionary is not None:
             self._set_read_dictionary(read_dictionary, &arrow_props)
 
+        if arrow_extensions_enabled:
+            arrow_props.set_arrow_extensions_enabled(<c_bool>True)
+
         with nogil:
             check_status(builder.memory_pool(self.pool)
                          .properties(arrow_props)
```

**Comment by pitrou** [comment]
@rok, yes, we should open a new issue for it

File: python/pyarrow/tests/test_extension_type.py
Comment (pitrou) [reply python/pyarrow/tests/test_extension_type.py:1979]: Indeed. `TypeError` would have to be raised at the C++ level instead. Anyway, this is out of scope for this PR.

**Review by pitrou — APPROVED** [review]

**Comment by rok** [comment]
Opened an issue for the `arrow_extensions_enabled` https://github.com/apache/arrow/issues/44500 problem.

**Pull request merged**
